internal/handlers/common/findandmodify.go

```diff
@@ -168,3 +178,152 @@ func GetFindAndModifyParams(doc *types.Document, l *zap.Logger) (*FindAndModifyP
 		MaxTimeMS:          maxTimeMS,
 	}, nil
 }
+
+// UpsertDocument updates the first document if exists, or create an insert document
+// from the params if no documents in query result docs.
+func UpsertDocument(docs []*types.Document, params *FindAndModifyParams) (*UpsertParams, error) {
+	res := new(UpsertParams)
+	var err error
+
+	if len(docs) == 0 {
+		res.Insert, err = insertDocuments(params)
+
+		res.ReturnValue = types.Null
+		if params.ReturnNewDocument {
+			res.ReturnValue = res.Insert
+		}
+
+		return res, err
+	}
+
+	res.Update, err = updateDocuments(docs, params)
+
+	res.ReturnValue = docs[0]
+	if params.ReturnNewDocument {
+		res.ReturnValue = res.Update
+	}
+
+	return res, err
+}
+
+// insertDocuments creates an insert document from the parameter.
+// When inserting new document we must check that `_id` is present, so we must extract `_id`
+// from query or generate a new one.
+func insertDocuments(params *FindAndModifyParams) (*types.Document, error) {
+	insert := must.NotFail(types.NewDocument())
+
+	if params.HasUpdateOperators {
+		if _, err := UpdateDocument(insert, params.Update); err != nil {
+			return nil, err
+		}
+	} else {
+		insert = params.Update
+	}
+
+	if !insert.Has("_id") {
+		id, err := getUpsertID(params.Query)
+		if err != nil {
+			return nil, err
+		}
+
+		insert.Set("_id", id)
+	}
+
+	return insert, nil
+}
+
+// updateDocuments updates the first document with update parameters.
+func updateDocuments(docs []*types.Document, params *FindAndModifyParams) (*types.Document, error) {
```

Comment:
Same here: this function doesn't update but prepares a document to be sent for update, right?

Reply:
Thanks, renamed to `prepareDocumentForInsert`